File: app/apis/vpic.py
import requests
from pydantic import ValidationError
from .exceptions import ApiError
from ..schemas.vin import Vin

class VpicApiError(ApiError):
  pass

def find_vin(vin: str) -> Vin | None:
  if not Vin.is_vin_correct_format(vin):
    raise ValueError(f'VIN {vin} must be a 17 alphanumeric characters string.')

  response = requests.get(
    f'https://vpic.nhtsa.dot.gov/api/vehicles/DecodeVinValues/{vin}?format=json',
    timeout=3
  )

  try:
    response.raise_for_status()
  except requests.HTTPError as ex:
    raise VpicApiError(f'Failed to get VIN {vin} from vpic API.', response.status_code) from ex

  json_obj = response.json()['Results'][0]
  try:
    return Vin(vin=vin,
               make=json_obj['Make'],
               model=json_obj['Model'],
               model_year=json_obj['ModelYear'],
               body_class=json_obj['BodyClass'])
  except ValidationError:
    return None

File: nodejs/src/apis/vpic.ts
import { InvalidArgumentError } from '../errors/argument-error';
import { Vin, isVinCorrectFormat, validateVin } from '../schemas/vin';

export class VpicApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CarImageryApiError';
  }
}

export async function findVin(vinNumber: string): Promise<Vin | null> {
  if (!isVinCorrectFormat(vinNumber)) {
    throw new InvalidArgumentError(`VIN ${vinNumber} must be a 17 alphanumeric characters string.`)
  }

  const url = `https://vpic.nhtsa.dot.gov/api/vehicles/DecodeVinValues/${vinNumber}?format=json`;
  const response = await fetch(url, { method: 'GET' });

  if (!response.ok) {
    throw new VpicApiError(`Failed to get VIN ${vinNumber} from vpic API. Status ${response.status} received.`);
  }

  const payload = (await response.json())['Results'][0];
  try {
    return validateVin({
      vinNumber: vinNumber,
      make: payload['Make'],
      model: payload['Model'],
      modelYear: payload['ModelYear'],
      bodyClass: payload['BodyClass'],
    });
  } catch (error) {
    return null;
  }
}
